src/lib/logger/impl.ml

```diff
@@ -0,0 +1,286 @@
+
+      type t =
+        { directory: string
+        ; max_size: int
+        ; mutable primary_log: File_descr.t
+        ; mutable primary_log_size: int }
+
+      let create ~directory ~max_size =
+        let primary_log_loc = Filename.concat directory primary_log_name in
+        let primary_log_size, mode =
+          if Result.is_ok (access primary_log_loc [`Exists; `Read; `Write])
+          then
+            let log_stats = stat primary_log_loc in
+            (Int64.to_int_exn log_stats.st_size, [O_RDWR; O_APPEND])
+          else (0, [O_RDWR; O_CREAT])
+        in
+        let primary_log = openfile ~perm:log_perm ~mode primary_log_loc in
+        {directory; max_size; primary_log; primary_log_size}
+
+      let rotate t =
+        let primary_log_loc = Filename.concat t.directory primary_log_name in
+        let secondary_log_loc =
+          Filename.concat t.directory secondary_log_name
+        in
+        close t.primary_log ;
+        rename ~src:primary_log_loc ~dst:secondary_log_loc ;
+        t.primary_log
+        <- openfile ~perm:log_perm ~mode:[O_RDWR; O_CREAT] primary_log_loc ;
+        t.primary_log_size <- 0
+
+      let transport t str =
+        if t.primary_log_size > t.max_size then rotate t ;
+        let len = String.length str in
+        if write t.primary_log ~buf:(Bytes.of_string str) ~len <> len then
+          printf "unexpected error writing to persistent log" ;
+        t.primary_log_size <- t.primary_log_size + len
+    end
+
+    let dumb_logrotate ~directory ~max_size =
+      T ((module Dumb_logrotate), Dumb_logrotate.create ~directory ~max_size)
+  end
+end
+
+module Consumer_registry = struct
+  type id = string
+
+  type consumer = {processor: Processor.t; transport: Transport.t}
+
+  type nonrec t = (id, consumer) List.Assoc.t ref
+
+  let t : t = ref []
+
+  let is_registered id = Option.is_some (List.Assoc.find !t id ~equal:( = ))
+
+  let register ~id ~processor ~transport =
+    if is_registered id then
+      failwith "cannot register logger consumer with the same id twice"
+    else t := List.Assoc.add !t id {processor; transport} ~equal:( = )
+
+  let broadcast_log_message msg =
+    List.iter !t ~f:(fun (_, consumer) ->
+        let (Processor.T ((module Processor_mod), processor)) =
+          consumer.processor
+        in
+        let (Transport.T ((module Transport_mod), transport)) =
+          consumer.transport
+        in
+        match Processor_mod.process processor msg with
+        | Some str ->
+            Transport_mod.transport transport str
+        | None ->
+            () )
+end
+
+type t = {null: bool; metadata: Metadata.t}
+
+let create ?(metadata = []) ?(initialize_default_consumer = true) () =
+  if initialize_default_consumer then
+    if not (Consumer_registry.is_registered "default") then
+      Consumer_registry.register ~id:"default" ~processor:(Processor.raw ())
+        ~transport:(Transport.stdout ()) ;
+  let pid = lazy (Unix.getpid () |> Pid.to_int) in
```

Comment:
Wait, why is this lazy? It's immediately forced.

Reply:
Good question... not sure exactly, I didn't add this line this time.